stdlib/public/core/Sort.swift.gyb

```diff
@@ -72,42 +72,87 @@ func _insertionSort<C>(
+  ${", by areInIncreasingOrder: (C.Iterator.Element, C.Iterator.Element) -> Bool" if p else ""}
+)
+  where
+  C : MutableCollection & RandomAccessCollection
+  ${"" if p else ", C.Iterator.Element : Comparable"}
+{
+  if !${cmp("elements[b]", "elements[a]", p)} {     // 123, 132, 231
+    if !${cmp("elements[c]", "elements[b]", p)} {   // 123
+      return
+    }
+
+    swap(&elements[b], &elements[c])                // 132->123, 231->213
+    if ${cmp("elements[b]", "elements[a]", p)} {    // 213
+      swap(&elements[a], &elements[b])              // 213->123
+    }
+    return
+  }
+                                                    // 213, 312, 321
+  if ${cmp("elements[c]", "elements[b]", p)} {      // 321
+    swap(&elements[a], &elements[c])                // 321->123
+    return
+  }
+
+  swap(&elements[a], &elements[b])                  // 213->123, 312->132
+  if ${cmp("elements[c]", "elements[b]", p)} {      // 132
+    swap(&elements[b], &elements[c])                // 132->123
+  }
+}
+
+/// Reorders `elements` and returns an index `p` such that every element in
+/// `elements[range.lowerBound..<pivot]` is less than every element in
+/// `elements[pivot..<range.upperBound]`.
+///
+/// - Precondition: The count of `range` must be >= 3:
+///   `elements.distance(from: range.lowerBound, to: range.upperBound) >= 3`
 func _partition<C>(
   _ elements: inout C,
   subRange range: Range<C.Index>
   ${", by areInIncreasingOrder: (C.Iterator.Element, C.Iterator.Element) -> Bool" if p else ""}
 ) -> C.Index
   where
   C : MutableCollection & RandomAccessCollection
-  ${"" if p else ", C.Iterator.Element : Comparable"} {
-
+  ${"" if p else ", C.Iterator.Element : Comparable"}
+{
   var lo = range.lowerBound
-  var hi = range.upperBound
+  var hi = elements.index(before: range.upperBound)
 
-  if lo == hi {
-    return lo
-  }
+  // Sort the first, middle, and last elements, then use the middle value
+  // as the pivot for the partition.
+  let half = elements.distance(from: lo, to: hi) / 2
+  let mid = elements.index(lo, offsetBy: half)
+  _sort3(&elements, lo, mid, hi
+    ${", by: areInIncreasingOrder" if p else ""})
+  let pivot = elements[mid]
 
-  // The first element is the pivot.
-  let pivot = elements[range.lowerBound]
+  /// The first and last elements are already in the correct partitions.
+  elements.formIndex(after: &lo)
+  elements.formIndex(before: &hi)
 
   // Loop invariants:
   // * lo < hi
-  // * elements[i] < pivot, for i in range.lowerBound+1..lo
-  // * pivot <= elements[i] for i in hi..range.upperBound
-
-Loop: while true {
-  FindLo: repeat {
-      elements.formIndex(after: &lo)
+  // * elements[i] < pivot, for i in range.lowerBound..<lo
+  // * pivot <= elements[i] for i in hi..<range.upperBound
+  Loop: while true {
+    FindLo: repeat {
       while lo != hi {
         if !${cmp("elements[lo]", "pivot", p)} { break FindLo }
         elements.formIndex(after: &lo)
       }
       break Loop
     } while false
```

Comment:
I'm guessing this code predates introduction of `do`-`catch`—also, I didn't realize you could break from a `do` block. I've switched that, thanks!